libs/s25main/mapGenerator/Resources.cpp

```diff
@@ -177,39 +178,73 @@ namespace rttr { namespace mapGenerator {
     {
         const auto& textures = map.textureMap;
         auto& resources = map.resources;
+        struct MineableResourceInfo
+        {
+            // The current "budget", that is, how many many are we over, or under, compared to the desired distribution.
+            int budget = 0;
+            // The ratio is the user defined ratio for this resource.
+            int ratio;
+            // The resource encoding
+            libsiedler2::Resource resource;
+            MineableResourceInfo(int ratio, libsiedler2::Resource resource) : ratio(ratio), resource(resource) {}
+        };
+        MineableResourceInfo mRIs[4] = {MineableResourceInfo(settings.ratioCoal, libsiedler2::R_Coal),
+                                        MineableResourceInfo(settings.ratioGold, libsiedler2::R_Gold),
+                                        MineableResourceInfo(settings.ratioIron, libsiedler2::R_Iron),
+                                        MineableResourceInfo(settings.ratioGranite, libsiedler2::R_Granite)};
         int total = settings.ratioCoal + settings.ratioGold + settings.ratioIron + settings.ratioGranite;
 
+        // Helper to pick an index [0,4) into the mRIs array that identifies a randomly selected resource we should
+        // place. Note that we do not perform budget adjustments and checks here.
+        auto resourcePicker = [&]() {
+            int ratio = 0;
+            int randomNumber = rnd.RandomValue(1, std::max(1, total));
+            for(int i = 0; i < 4; ++i)
+            {
+                ratio += mRIs[i].ratio;
+                if(randomNumber < ratio)
+                    return i;
+            }
+            return -1;
```

Comment:
This should never happen, should it? But it makes me think: What if all ratios are 0? Then no resources should be placed, right? Maybe simply assert that here and never call the function for that (place the `if(total==0)` in the loop below) or check that first in the function and return -1 right away. Either way you can then also remove  the `max` call.

As you have to return something, the loop can go to `mRIs.size() - 1` and you return that here, no need to add and check the ratio for the last element anyway as it will always be true after the change to `<=` above

Reply:
This is for the 0 case. I can rewrite it obviously...